### ISSUE TITLE
JASP has difficulty handling extended missing value codes in Stata

### ISSUE THREAD

**Issue by jlwoodard** [jasp-stats/jasp-issues#733]
* JASP version: 0.12.2
 MacOS
* OS name and version: Version 10.15.4, Catalina
* Analysis: Loading a data file created by Stata into JASP
* Bug description: If the data file contains extended missing value codes (such as .a, .m, .p, etc., that reflect not only that the data are missing but why the data are missing), the extended missing values seem to be interpreted as actual data values. There is an integer that displays in the data editor of JASP and in a descriptive statistics frequency table analysis that seems to correspond to the extended missing value code. Computation of means and standard deviations seem to interpret these integers as actual numbers (rather than as missing) and create an error in any analyses using a variable with extended missing value codes.
* Expected behaviour: I would expect that the extended missing value codes would be recognized as missing. If they cannot retain their extended codes, if they can at least be imported as missing data, that would be a huge help.

#### Steps to reproduce:
1. Go to 'Open JASP, open the Stata data file. Look at the JASP data editor. See triple digit values or occasionally the word "none" in many of the cells in the attached data file'
2. Click on 'Descriptive Statistics and perform descriptive analysis on the variables in the attached data file. Also check frequency tables to reveal triple digit values that seem to correspond to a particular type of extended missing value. Also note the means and standard deviations are including the triple digit missing values. For the attached data, the maximum score for var1 should be 22, and the maximum score for var2 and var3 should be only 10'

4. I have included the Stata data file so you can see how Stata codes the extended missing values and after importing it to JASP, you can see how they are interpreted during the input.
[Stata Sample File.dta.zip](https://github.com/jasp-stats/jasp-issues/files/4583658/Stata.Sample.File.dta.zip)

**Comment by boutinb** [comment]
@jlwoodard If you click on the column header in the Data panel of JASP, then you can see the values and labels read from the file. You can see for example for var3, that 0 corresponds to label 'none', as long as values 1, 2 or even 110 and 114.
If in JASP, you add the missing values (in Preferences/Data) 0, 110 and 114, you will see that these values will automatically be removed form the Data panel, and that in Description they will be considered as missing values.
This is of course a workaround, and we have to understand why JASP does not read automatically from the Stata file what are the missing values.

**Comment by JorisGoosen** [comment]
@jlwoodard thank you for creating and including the stata datafile. Right now I don't have access to stata itself, although I might buy it to debug this. But for now you could help me out perhaps? 
This can be done by describing the missing values so that I can debug the loading of them in JASP.

**Comment by jlwoodard** [comment]
@boutinb Thanks so much for your suggestion! I actually started doing what you suggested earlier, but it occurred to me that values like 0 (none) was a valid data value, so I didn't want to exclude it. I think for that particular value, I would need to convert the 0 (none) to a true integer rather than a value label from within Stata. The other two (110 and 114) would definitely be missing in this small dataset, but in the larger dataset, making them missing within JASP might exclude valid values of 110 and 114 for other variables. As you note though, it is a good workaround for variables where those 3 digit numbers are not valid data.

**Comment by jlwoodard** [comment]
@JorisGoosen Thank you so much for trying to debug this problem. Stata may have a free trial that may let you access the program if needed for debugging, but I'll certainly help as much as possible. I've attached a missing value analysis from this particular small dataset, and within Stata, the extended missing value codes appear to be .i, .m, .n, .r, and .s.  In other data sets, the extended missing value codes could be a period plus any letter. Within Stata, .i might appear as a word (inappropriate), .n might appear in the Stata data editor as "not ascertained", etc. In this data set, var1 has a maximum valid value of 22, so any integer greater than 22 probably reflects a missing value code that is being translated from the Stata extended value to something else within JASP. Variables 2 and 3 have maximum valid values of 10. As best as I can tell, .i is being translated to 110, .m is being translated to 114, .n is being translated to 115, .s is being translated to 120, and .r is being translated to 119. It seems pretty systematic that JASP is numbering the letter associated with the MV code according to its alphabetical position. I hope that description of the missing data is helpful in enabling you to debug this issue. Let me know if you need anything else, and many thanks in advance!  John
[Screen Shot 2020-05-06 at 10.28.20.zip](https://github.com/jasp-stats/jasp-issues/files/4587440/Screen.Shot.2020-05-06.at.10.28.20.zip)

**Comment by JorisGoosen** [comment]
Well, Im hoping I can get it with some education discount, maybe the price isn't so bad then.
I already noticed it was next to impossible to debug problems in loading SPSS files. Im guessing the same goes for Stata (and SAS...).

I'll have a look soon and thanks for all the info!

**Comment by JorisGoosen** [comment]
@jlwoodard so I've had a look at this problem and did notice we only checked for so called "system missing values" and were indeed missing the "tagged missing values". 
With some changes this is now taken into account. Some values above 22 (110, 114, 115 and 119 to be specific) are now interpreted as missing.

Because of some limitations on how we store missing values inside JASP (globally for the application I think, but I need to verify this) I have now made it so it drops these values.

Are these the incorrect values you saw popping up?

I did notice the following in the labels:
![image](https://user-images.githubusercontent.com/29062713/82813506-7fd1e800-9e95-11ea-8c1c-9dd1471224cb.png)
And Im wondering if those labels at the bottom with values >2000000000 are those corresponding to the missing values that were just dropped?

Because I looked at licensing options for STATA and the prices are positively outrageous from my perspective...

**Comment by JorisGoosen** [comment]
And I just heard from a colleague that the missing values *are* stored per jasp-file, so maybe I can process them without dropping them. I will add an issue for this in our private issue tracker. But at the very least your missing values will no longer be interpreted as actual values.

**Comment by jlwoodard** [comment]
Hi Joris,
   Thanks so much for getting back to me! Yes, the values in the table you sent with text in the labels are the “extended system missing values” that were coded by Stata as missing. Let me know if I can provide more information or if you would like me to test a version of JASP or to send you a new data file.

Many thanks again!

John
On May 25, 2020, 08:40 -0400, Joris Goosen <notifications@github.com>, wrote:

@jlwoodard<https://github.com/jlwoodard> so I've had a look at this problem and did notice we only checked for so called "system missing values" and were indeed missing the "tagged missing values".
With some changes this is now taken into account. Some values above 22 (110, 114, 115 and 119 to be specific) are now interpreted as missing.

Because of some limitations on how we store missing values inside JASP (globally for the application I think, but I need to verify this) I have now made it so it drops these values.

Are these the incorrect values you saw popping up?

I did notice the following in the labels:
[image]<https://user-images.githubusercontent.com/29062713/82813506-7fd1e800-9e95-11ea-8c1c-9dd1471224cb.png>
And Im wondering if those labels at the bottom with values >2000000000 are those corresponding to the missing values that were just dropped?

Because I looked at licensing options for STATA and the prices are positively outrageous from my perspective...
